src/Components/test/testassets/TestContentPackage/Services/InteractiveWebAssemblyService.cs

```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.Components;
+
+namespace TestContentPackage.Services;
+
+public class InteractiveWebAssemblyService
+{
+    [SupplyParameterFromPersistentComponentState]
+    public string State { get; set; }
```

Comment:
Does it work with private property?
Having a public property on service class doesn't seem much encapsulated.

Reply:
Your class might be internal, but the property needs to be public I believe.

It's customary across different features, especially those involving serialization in ASP.NET that we require the properties for things to be public. It's also important if in the future we decide to use a different approach other than reflection (like source generation).

Reply:
Possible source generation argument is great. On the other hand having a public property on a service that is public just because of the implementation detail seems wrong, although it can designed in a different way if that's a concern.

Reply:
> On the other hand having a public property on a service that is public just because of the implementation detail seems wrong, although it can designed in a different way if that's a concern.

If your type is `internal` it doesn't really matter. You can still use the imperative API if you really don't want to expose anything even inside the assembly itself.